lib/std/http/Client.zig

```diff
@@ -1,62 +1,447 @@
+//! open connections to skip DNS and TLS handshake for subsequent requests.
 
 const std = @import("../std.zig");
+const mem = std.mem;
 const assert = std.debug.assert;
 const http = std.http;
 const net = std.net;
 const Client = @This();
 const Url = std.Url;
+const Allocator = std.mem.Allocator;
+const testing = std.testing;
 
-allocator: std.mem.Allocator,
-headers: std.ArrayListUnmanaged(u8) = .{},
-active_requests: usize = 0,
+/// Used for tcpConnectToHost and storing HTTP headers when an externally
+/// managed buffer is not provided.
+allocator: Allocator,
 ca_bundle: std.crypto.Certificate.Bundle = .{},
 
+pub const Connection = struct {
+    stream: net.Stream,
+    /// undefined unless protocol is tls.
+    tls_client: std.crypto.tls.Client,
+    protocol: Protocol,
+
+    pub const Protocol = enum { plain, tls };
+
+    pub fn read(conn: *Connection, buffer: []u8) !usize {
+        switch (conn.protocol) {
+            .plain => return conn.stream.read(buffer),
+            .tls => return conn.tls_client.read(conn.stream, buffer),
+        }
+    }
+
+    pub fn readAtLeast(conn: *Connection, buffer: []u8, len: usize) !usize {
+        switch (conn.protocol) {
+            .plain => return conn.stream.readAtLeast(buffer, len),
+            .tls => return conn.tls_client.readAtLeast(conn.stream, buffer, len),
+        }
+    }
+
+    pub fn writeAll(conn: *Connection, buffer: []const u8) !void {
+        switch (conn.protocol) {
+            .plain => return conn.stream.writeAll(buffer),
+            .tls => return conn.tls_client.writeAll(conn.stream, buffer),
+        }
+    }
+
+    pub fn write(conn: *Connection, buffer: []const u8) !usize {
+        switch (conn.protocol) {
+            .plain => return conn.stream.write(buffer),
+            .tls => return conn.tls_client.write(conn.stream, buffer),
+        }
+    }
+};
+
 /// TODO: emit error.UnexpectedEndOfStream or something like that when the read
 /// data does not match the content length. This is necessary since HTTPS disables
 /// close_notify protection on underlying TLS streams.
 pub const Request = struct {
     client: *Client,
-    stream: net.Stream,
-    headers: std.ArrayListUnmanaged(u8) = .{},
-    tls_client: std.crypto.tls.Client,
-    protocol: Protocol,
-    response_headers: http.Headers = .{},
+    connection: Connection,
+    redirects_left: u32,
+    response: Response,
+    /// These are stored in Request so that they are available when following
+    /// redirects.
+    headers: Headers,
 
-    pub const Protocol = enum { http, https };
+    pub const Response = struct {
+        headers: Response.Headers,
+        state: State,
+        header_bytes_owned: bool,
+        /// This could either be a fixed buffer provided by the API user or it
+        /// could be our own array list.
+        header_bytes: std.ArrayListUnmanaged(u8),
+        max_header_bytes: usize,
 
-    pub const Options = struct {
+        pub const Headers = struct {
```

Comment:
is the path of this `std.http.Request.Response.Headers` ?

Reply:
`std.http.Client.Request.Response.Headers`

Reply:
can `Request`/`Response`/`Headers` be moved to `std.http` top level?